File: packages/compliance/src/compliance/terminology.py
"""Presentation terminology — Research Mode vs SEBI Mode (PR1.0).

Maps engine/domain action tokens to **user-facing labels**.
Does not mutate recommendation or decision engines.
"""

from __future__ import annotations

from enum import StrEnum

from compliance.feature_flags import FeatureFlags

__all__ = [
    "ResearchLabel",
    "FIELD_LABELS_RESEARCH",
    "FIELD_LABELS_SEBI",
    "present_action",
    "present_field_label",
]


class ResearchLabel(StrEnum):
    """Neutral DSP View labels used in Research Mode."""

    ATTRACTIVE = "Attractive"
    CAUTION = "Caution"
    FAIRLY_VALUED = "Fairly Valued"
    REDUCE_EXPOSURE = "Reduce Exposure"
    WATCH = "Watch Closely"
    INSUFFICIENT_EVIDENCE = "Insufficient Evidence"
    UNKNOWN = "Unclassified"


# Canonical engine / API tokens → Research Mode DSP View
_ACTION_TO_RESEARCH: dict[str, ResearchLabel] = {
    "strong_buy": ResearchLabel.ATTRACTIVE,
    "buy": ResearchLabel.ATTRACTIVE,
    "hold": ResearchLabel.FAIRLY_VALUED,
    "neutral": ResearchLabel.FAIRLY_VALUED,
    "reduce": ResearchLabel.REDUCE_EXPOSURE,
    "sell": ResearchLabel.CAUTION,
    "strong_sell": ResearchLabel.CAUTION,
    "watch": ResearchLabel.WATCH,
    "insufficient_evidence": ResearchLabel.INSUFFICIENT_EVIDENCE,
}

# SEBI / recommendation mode retains conventional labels (architecture only)
_ACTION_TO_SEBI: dict[str, str] = {
    "strong_buy": "Strong Buy",
    "buy": "Buy",
    "hold": "Hold",
    "neutral": "Hold",
    "reduce": "Reduce",
    "sell": "Sell",
    "strong_sell": "Strong Sell",
    "watch": "Watch",
    "insufficient_evidence": "Insufficient Evidence",
}

FIELD_LABELS_RESEARCH: dict[str, str] = {
    "target_price": "Estimated Intrinsic Value Range",
    "recommendation": "Research Conclusion",
    "stock_recommendation": "Investment Assessment",
    "action": "DSP View",
    "official_target_price": "Estimated Intrinsic Value Range",
}

FIELD_LABELS_SEBI: dict[str, str] = {
    "target_price": "Official Target Price",
    "recommendation": "Recommendation",
    "stock_recommendation": "Stock Recommendation",
    "action": "Recommendation",
    "official_target_price": "Official Target Price",
}


def present_action(token: str, flags: FeatureFlags | None = None) -> str:
    """Return user-facing action / posture label for ``token``.

    Parameters
    ----------
    token:
        Engine or API action string (e.g. ``buy``, ``STRONG_BUY``).
    flags:
        When ``allow_action_labels()`` is true, SEBI wording is used.
        Otherwise Research Mode neutral DSP View labels are used.
    """
    key = token.strip().lower().replace(" ", "_").replace("-", "_")
    active = flags or FeatureFlags()
    if active.allow_action_labels():
        return _ACTION_TO_SEBI.get(key, token)
    label = _ACTION_TO_RESEARCH.get(key, ResearchLabel.UNKNOWN)
    return str(label)


def present_field_label(field_key: str, flags: FeatureFlags | None = None) -> str:
    """Return user-facing field title for a conceptual field key."""
    key = field_key.strip().lower()
    active = flags or FeatureFlags()
    table = (
        FIELD_LABELS_SEBI
        if active.allow_official_target_price() or active.allow_action_labels()
        else FIELD_LABELS_RESEARCH
    )
    if active.allow_official_target_price() and key in {
        "target_price",
        "official_target_price",
    }:
        return FIELD_LABELS_SEBI["official_target_price"]
    if not active.allow_action_labels() and key in FIELD_LABELS_RESEARCH:
        return FIELD_LABELS_RESEARCH[key]
    return table.get(key, field_key)

File: apps/web/src/lib/terminology.ts
/** Research Mode terminology — never hard-code BUY/SELL/HOLD in UI. */

import {
  allowActionLabels,
  allowOfficialTargetPrice,
  featureFlags,
  type FeatureFlags,
} from "@/lib/featureFlags";

const ACTION_RESEARCH: Record<string, string> = {
  strong_buy: "Attractive",
  buy: "Attractive",
  hold: "Fairly Valued",
  neutral: "Fairly Valued",
  reduce: "Reduce Exposure",
  sell: "Caution",
  strong_sell: "Caution",
  watch: "Watch Closely",
  insufficient_evidence: "Insufficient Evidence",
};

const ACTION_SEBI: Record<string, string> = {
  strong_buy: "Strong Buy",
  buy: "Buy",
  hold: "Hold",
  neutral: "Hold",
  reduce: "Reduce",
  sell: "Sell",
  strong_sell: "Strong Sell",
  watch: "Watch",
  insufficient_evidence: "Insufficient Evidence",
};

const FIELD_RESEARCH: Record<string, string> = {
  target_price: "Estimated Intrinsic Value Range",
  recommendation: "Research Conclusion",
  stock_recommendation: "Investment Assessment",
  action: "DSP View",
};

const FIELD_SEBI: Record<string, string> = {
  target_price: "Official Target Price",
  recommendation: "Recommendation",
  stock_recommendation: "Stock Recommendation",
  action: "Recommendation",
};

function normalizeToken(token: string): string {
  return token.trim().toLowerCase().replace(/[\s-]+/g, "_");
}

export function presentAction(
  token: string,
  flags: FeatureFlags = featureFlags,
): string {
  const key = normalizeToken(token);
  if (allowActionLabels(flags)) {
    return ACTION_SEBI[key] ?? token;
  }
  return ACTION_RESEARCH[key] ?? "Unclassified";
}

export function presentFieldLabel(
  fieldKey: string,
  flags: FeatureFlags = featureFlags,
): string {
  const key = normalizeToken(fieldKey);
  if (allowOfficialTargetPrice(flags) && key === "target_price") {
    return FIELD_SEBI.target_price;
  }
  if (allowActionLabels(flags)) {
    return FIELD_SEBI[key] ?? fieldKey;
  }
  return FIELD_RESEARCH[key] ?? fieldKey;
}
